app/controllers/products_controller.rb

```diff
@@ -48,6 +48,11 @@ def destroy
   end
 
   def add_to_cart
+    unless logged_in?
```

Comment:
You can extract this in a before_action, since it's an action that will likely be common to other controllers too, you can define it in ApplicationController, as the other controllers extend it.